torch/autograd/graph.py

```diff
@@ -269,3 +271,125 @@ def __setstate__(self, state):
+
+def _get_sid(t) -> Tuple[int, int]:
+    return (t.data_ptr(), t._version)
+
+class _Handle():
+    pass
+
+class _swap_with_cloned(saved_tensors_hooks):
+    def __init__(self):
+        def pack_hook(t):
+            tid = _get_tid(t)
+            sid = _get_sid(t)
+            # Tensors saved for backward have an entry in _tid_to_weakhandle
+            handle: Optional[_Handle] = None
+
+            # Save aliasing information
+            if sid not in _sid_to_tid:
+                _sid_to_tid[sid] = set()
+            _sid_to_tid[sid].add(tid)
+
+            # NB: The same tensor (of the same version) can be saved multiple times
+            if tid not in _tid_to_weakhandle or _tid_to_weakhandle[tid]() is None:
+                handle = _Handle()
+                _tid_to_weakhandle[tid] = weakref.ref(handle)
+                _original[handle] = t
+            else:
+                # Store an additional handle
+                handle = _tid_to_weakhandle[tid]()
+            return _ctx_id, handle
+
+        def unpack_hook(tup):
+            ctx_id, handle = tup
+            assert ctx_id == _ctx_id, (
+                "Trying to backward outside of the 'allow_mutation_on_saved_tensors' context"
+                "in which the graph was originally recorded.")
+            if handle in _cloned:
+                res = _cloned[handle]
+            else:
+                res = _original[handle]
+            return res
+
+        super().__init__(pack_hook, unpack_hook)
+
+class _CloneArgBeforeMutateMode(TorchDispatchMode):
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+
+        # Bug? Maybe something to do with wrapped number + dispatcher, not sure why yet
+        # We don't really need this unless we try to print though, but keeping for now
+        if type(args[0]) is int:
+            args = [torch.tensor(args[0]), *args[1:]]
+
+        # (only for in-place ops now, we may want to handle out= later)
+        if func.__name__.split('.')[0][-1] == "_":
+            # The first argument is assumed to be modified in-place
+            tid = _get_tid(args[0])
+            sid = _get_sid(args[0])
+            if sid in _sid_to_tid:
+                for tid in _sid_to_tid[sid]:
+                    if tid not in _tid_to_weakhandle:
+                        # It's never been saved
+                        continue
+                    handle = _tid_to_weakhandle[tid]()
+                    if handle is None or handle in _cloned:
+                        # It's been saved, but backward was run OR
+                        # The same exactly tensor has been cloned already
+                        continue
+                    _cloned[handle] = _original[handle].clone()
+                    del _original[handle]
+            else:
+                # this can happen with math views, I'm not sure why yet
+                assert not args[0]._is_view()
```

Comment:
Why this assert? You forbid inplace on views?

Reply:
uhh this might've just been code to debug something with math bits. This was to show that it WASN'T going down this path even though I WAS doing inplace on view, so the conj bit must've been materialized somewhere

There's should be no need for special handling of math bits with this implementation so I can probably remove the assert - I just wanted to write a test and show that we do go doing this path.